src/coverlet.core/Instrumentation/ModuleTrackerTemplate.cs

```diff
@@ -69,77 +73,90 @@ public static void UnloadModule(object sender, EventArgs e)
-                    using (var br = new BinaryReader(fs))
-                    using (var bw = new BinaryWriter(fs))
+                    catch (PlatformNotSupportedException)
                     {
-                        int hitsLength = br.ReadInt32();
-                        if (hitsLength != HitsArray.Length)
-                        {
-                            throw new InvalidOperationException(
-                                $"{HitsFilePath} has {hitsLength} entries but on memory {nameof(HitsArray)} has {HitsArray.Length}");
-                        }
+                        memoryMap = MemoryMappedFile.CreateFromFile(HitsFilePath, FileMode.Open, null, (HitsArraySize + Coverage.HitsResultHeaderSize) * sizeof(int));
+                    }
 
-                        for (int i = 0; i < hitsLength; ++i)
+                    // Tally hit counts from all threads in memory mapped area
+                    var accessor = memoryMap.CreateViewAccessor();
+                    using (var buffer = accessor.SafeMemoryMappedViewHandle)
+                    {
+                        unsafe
                         {
-                            int oldHitCount = br.ReadInt32();
-                            bw.Seek(-sizeof(int), SeekOrigin.Current);
-                            bw.Write(HitsArray[i] + oldHitCount);
+                            byte* pointer = null;
+                            buffer.AcquirePointer(ref pointer);
+                            try
+                            {
+                                var intPointer = (int*) pointer;
+
+                                // Signal back to coverage analysis that we've started transferring hit counts.
+                                // Use interlocked here to ensure a memory barrier before the Coverage class reads
+                                // the shared data.
+                                Interlocked.Increment(ref *(intPointer + Coverage.HitsResultUnloadStarted));
+
+                                for (var i = 0; i < HitsArraySize; i++)
+                                {
+                                    var count = 0;
+
+                                    foreach (var threadHits in threads)
+                                    {
+                                        count += threadHits[i];
+                                    }
+
+                                    if (count > 0)
+                                    {
+                                        // There's a header of one int before the hit counts
+                                        var hitLocationArrayOffset = intPointer + i + Coverage.HitsResultHeaderSize;
+
+                                        // No need to use Interlocked here since the mutex ensures only one thread updates 
+                                        // the shared memory map.
+                                        *hitLocationArrayOffset += count;
```

Comment:
❔ Is the file guaranteed to be zero-filled when it is first created?

Reply:
Yes, it's either a pure memory map (on Windows) or a new and thus empty file (on Unixlike), and in both cases zero pages are allocated for the map.